transports/quic/src/connection/stream.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2022 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{
+    io::{self},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{AsyncRead, AsyncWrite};
+
+/// A single stream on a connection
+pub struct Stream {
+    /// A send part of the stream
+    send: quinn::SendStream,
+    /// A receive part of the stream
+    recv: quinn::RecvStream,
+    /// Whether the stream is closed or not
+    close_result: Option<Result<(), ()>>,
+}
+
+impl Stream {
+    pub(super) fn new(send: quinn::SendStream, recv: quinn::RecvStream) -> Self {
+        Self {
+            send,
+            recv,
+            close_result: None,
+        }
+    }
+}
+
+impl AsyncRead for Stream {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        if let Some(close_result) = self.close_result {
+            if close_result.is_err() {
+                return Poll::Ready(Ok(0));
+            }
+        }
+        let read_result = futures::ready!(Pin::new(&mut self.recv).poll_read(cx, buf));
+        if read_result.is_err() {
+            return Poll::Ready(Ok(0));
+        }
+        Poll::Ready(read_result)
+    }
+}
+
+impl AsyncWrite for Stream {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        Pin::new(&mut self.send).poll_write(cx, buf)
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>> {
+        Pin::new(&mut self.send).poll_flush(cx)
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>> {
+        if let Some(close_result) = self.close_result {
+            // For some reason poll_close needs to be 'fuse'able
+            return Poll::Ready(close_result.map_err(|()| io::ErrorKind::Unsupported.into()));
```

Comment:
`io::ErrorKind` is `Clone` and `Copy` so you could extract that from the original error and create the same. You'll lose any attached `source` errors but at least we return the same kind. What do you think?

Reply:
Done